### ISSUE TITLE
[C#] Implement new Concept Exercise: randomness

### ISSUE THREAD

**Issue by mikedamay** [exercism/v3#1642]
This issue describes how to implement the `regular-expressions` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student how to implement randomness in C#.

## Learning objectives

- Know how to implement randomness with `System.Random` in C#.

## Out of scope

- Secure random numbers

## Concepts

- `randomness` as they are handled in C# and .NET.

## Prerequisites

- `numbers`
- `strings`
-  `conditionals-if`

Other prerequisites will depend on the story.

### Hints
[random class][random-class] documentation shows how to use `System.Random`.

### After

[random class][random-class] documentation shows how to use `System.Random`.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[random-class]: https://docs.microsoft.com/en-us/dotnet/api/system.random?view=netcore-3.1
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer

**Comment by mikedamay** [comment]
I'm thinking of specifying the throw of a die in the context of the wizards and warriors theme.

**Comment by ErikSchierboom** [comment]
@mikedamay I think that would work really well as the context. Did you know that we have an existing exercise on randomness named [`dnd-character`](https://github.com/exercism/csharp/tree/master/exercises/dnd-character)? That might be useful for inspiration. If you feel that there is too much overlap, we could always consider retracting the D&D character exercise.

**Comment by mikedamay** [comment]
A number of students including me have had trouble understanding D&D character.   D&D passed me by somehow - more of a Risk or Diplomacy man.

I sent you the wrong link - [this](https://github.com/exercism/v3/pull/1659) would have been more useful.  It's a very stripped down puzzle and left to my devices I wasn't thinking of adding anything .

**Comment by ErikSchierboom** [comment]
> It's a very stripped down puzzle and left to my devices I wasn't thinking of adding anything .

I would consider it just a little _too_ basic to be honest. What if we could randomly generate names by choosing (and combining) from an array of first names and a list of last names (or nick names or whatever works). Or generate a random adversary with some random data?

**Comment by mikedamay** [comment]
> I would consider it just a little too basic

I would claim that the puzzle, as it stands, illustrates the important features of C#'s `System.Random`.
- it generates random numbers within a range
- it does not require a seed
- The upper bound of the overloaded methods that take 2 parameters is exclusive.

_after.md_ can provide the following:
- reinforce no requirement for a seed (and perhaps pre-empt confusion regarding similar implementations that do require one).
- show the use case for a seed and discuss determinism 
- mention the other generation overloads (in particular, reals).
- talk about its inapplicability to security


> we could randomly generate names

I concede that this teaches the student that once we have generated random numbers they can be used in many non-numeric situations but a) this is unlikely to be the student's first exposure to random numbers, b) practice exercises will bring this up and c) the cost is a lot of noise unrelated to the key lessons.

If we want to make the exercise more substantial then we could consider handling real numbers as well, say with some sort of _Spell Strength__ or something (I really don't know about D&D).

**Comment by ErikSchierboom** [comment]
> show the use case for a seed and discuss determinism

What if we can somehow incorporate this into the exercise? Maybe we could provide an additional method to generate a sequence of numbers? That way we could "detect" if the student just kept on creating new `Random` instances, which would lead to invalid results (likely). Speaking of which, this might be nice to do with `yield` too, generating an infinite sequence of random numbers.

**Comment by mikedamay** [comment]
> What if we can somehow incorporate this into the exercise?

This certainly overcomes the problem with testing correct usage.  I felt that the advantage we have here over exercises like _robot-name_ is that we can detect correct usage using the analyser or (as my proposed code is so simple that it's likely to have only a couple of representations) we  could even just rely on the representer mechanism.   

The problem as I'm sure you appreciate is that we are starting the student off with a very minor use case. 

The other problem is that the story may be artificial.  I suppose we could make it some sort of QA scenario when we are sampling something or other.  But we would still need to have the seedless version of `Random` taking some part and my imagination comes up short there.

Were you thinking of `yield` in this exercise?  I can see`yield` + `Random` being a good fit in some LINQ exercise but I would have thought we want to add `yield` as either a concept or a pre-requisite in this exercise unless you have a particular vision.

**Comment by ErikSchierboom** [comment]
> Were you thinking of yield in this exercise?

I was not sure, just wanted to put it out as an option,

> If we want to make the exercise more substantial then we could consider handling real numbers as well, say with some sort of Spell Strength_ or something (I really don't know about D&D).

That could work well. We can probably find a D&D related real number use case as well. Shall we then try to add that one extra method? I like the addition of returning a random double besides returning a random int.

**Comment by mikedamay** [comment]
closed by PR #1659
